Add the `Data Integration` tutorial (#25)
ballerina-guides/integration-tutorials: data-integration/README.md

# Data integration

## Overview

> In this tutorial, you will develop a service to manage data stored in a database using bal persist. The service will be used to create, retrieve, update and delete data from the database.

Reviewer: ```suggestion
In this tutorial, you will develop a service to manage data stored in a database using the Ballerina persistence layer. The service will be used to create, retrieve, update and delete data from the database.
```

I think we need to phrase these more as a service backed by a database, rather than exposing the database via a service.

> To implement this use case, you will develop a REST service with multiple resources using Visual Studio Code with the Ballerina Swan Lake extension and define the `data model` using the `bal persist` feature.

Reviewer: We emphasize (with backticks) only keywords, names, commands, etc. We shouldn't emphasize "data model".

> To implement this use case, you will develop a REST service with multiple resources using Visual Studio Code with the Ballerina Swan Lake extension and define the `data model` using the `bal persist` feature.

Reviewer: I would break the latter bit about persistence and usage to a separate sentence. Also mention working with the data/database in addition to defining the model.

The flow is as follows.

> 1. Create an employee.

Reviewer: I would include something like "Send a POST request to ..."

   ```json
   {
        "id": 1,
        "name": "John Doe",
        "age": 22,
        "phone": "8770586755",
        "email": "johndoe@gmail.com",
        "department": "IT"
   }

>    ```

Reviewer: ````suggestion
   ```

````

Reviewer: Check wherever applicable.

2. Create a task and assign it to the employee.

    ```json
    {
        "taskId": 1001,
        "description": "Analyze network performance",
        "status": "IN_PROGRESS",
        "employeeId": 1
    }
    ```
3. Update the status of the task.
   
    ```json
    {
        "description": "Analyze network performance",
        "status": "COMPLETED",
        "employeeId": 1
    }
    ```
4. Delete the task by giving the task ID of the completed task that needs to be deleted.

### Concepts covered

- REST API

> - Bal persist

Reviewer: ```suggestion
- bal persist
```

or

```suggestion
- The Ballerina persistence layer
```

- CRUD operations

## Develop the application

### Step 1: Set up the workspace

Install [Ballerina Swan Lake](https://ballerina.io/downloads/) and the [Ballerina Swan Lake extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina) on VS Code.

### Step 2: Develop the service

Follow the instructions given in this section to develop the service.

1. Create a new Ballerina project using the `bal` command and open it in VS Code.

    ```bash
    $ bal new data-integration
    ```

2. Initialize `bal persist` in the project. Specify the module name as `store` and the datastore as `mysql`. This will create a persist directory in the project root directory. The directory will contain the `model.bal` file, which is used to define the data model.

    ```bash
    $ bal persist init --module store --datastore mysql
    ```

3. Open the `model.bal` file in the persist directory and define the data model as follows.

> 

Reviewer: May make sense to explain the model (relationships) briefly.

    ```ballerina
    import ballerina/persist as _;

    type Employee record {|
        readonly int id;
        string name;
        int age;
        string phone;
        string email;
        string department;
        Task[] tasks;
    |};

    type Task record {|
        readonly int taskId;
        string taskName;
        string description;
        TaskStatus status;
        Employee employee;
    |};

    public enum TaskStatus {
        NOT_STARTED,
        IN_PROGRESS,
        COMPLETED
    };
    ```

    **Note:** 
    > The entities defined in the `model.bal` file should be based on the [`persist model specification`](https://ballerina.io/learn/persist-model/).

> 4. Generate the client objects, types and and scripts for the model using the following command. This will parse the `persist/model.bal` file and add the generated files under the generated directory..

Reviewer: Extra period at the end.

May make sense to include the module name.

   
    ```bash
    $ bal persist generate --module store
    ```

> 5. Remove the generated content in the `main.bal` file and open the diagram view in VS Code.

Reviewer: `main.bal` file in the project root

    ![Open diagram view](./resources/open_diagram_view.gif)

> 6. Define the generated client object `store:Client`.

Reviewer: ```suggestion
6. Define a variable of the generated client object `store:Client` type to interact with the database.
```

   
   ```ballerina
   import data_integration.store;

   final store:Client dbClient = check new;
   ```
   

> 7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resources that accepts user requests, interact with the database and manage operations to create, retrieve, update and delete employee and task data.

Reviewer: ```suggestion
7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resources that accept user requests, interact with the database, and manage operations to create, retrieve, update, and delete employee and task data.
```

>   - Open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.

Reviewer: Recheck indentation everywhere.

  - Define a service attached to the listener that is listening on port `9090`.

        ![Define the service](./resources/define_a_service.gif)

>   - Define HTTP resources that allow CRUD operations on the employee data. The following GIF shows how to define the `post` resource that creates an employee. Similarly, define the `get`, `put` and `delete` resources.

Reviewer: post -> POST

Please fix for everything.

        ![Define the resource](./resources/define_the_resource.gif)

8. Implement the logic.

```ballerina
service / on new http:Listener(9090) {
    resource function post employees(store:EmployeeInsert employee)
            returns http:Created|http:Conflict|http:InternalServerError {
        int[]|persist:Error result = dbClient->/employees.post([employee]);
        if result is persist:AlreadyExistsError {
            return http:CONFLICT;
        }
        if result is persist:Error {
            return <http:InternalServerError>{body: result.message()};
        }
        return http:CREATED;
    }
}
```

>    - The `post` resource accepts a `store:EmployeeInsert` object as the payload and returns a `http:Created` response if the employee is created successfully. If the employee already exists, it returns an `http:Conflict` response and if an error occurs, it returns an `http:InternalServerError` response.

Reviewer: object or record?

>    - The `post` resource accepts a `store:EmployeeInsert` object as the payload and returns a `http:Created` response if the employee is created successfully. If the employee already exists, it returns an `http:Conflict` response and if an error occurs, it returns an `http:InternalServerError` response.

Reviewer: a `http:Created` -> an `http:Created`

>    - Similarly, implement the logic for the other resources.

Reviewer: We decided to explain one where the client call is included in a query expression, right? To explain filtering/operations at db level?

        }
        return updatedTask;
    }

    resource function delete task/[int taskId]()
            returns http:NoContent|http:NotFound|http:InternalServerError {
        store:Task|persist:Error task = dbClient->/tasks/[taskId];
        if task is persist:NotFoundError {
            return http:NOT_FOUND;
        }
        if task is persist:Error {
            return <http:InternalServerError>{body: task.message()};
        }
        if task.status != store:COMPLETED {
            return <http:InternalServerError>{body: "Task is not completed yet"};
        }
        store:Task|persist:Error deleteResult = dbClient->/tasks/[taskId].delete;
        if deleteResult is persist:Error {
            return <http:InternalServerError>{body: deleteResult.message()};
        }
        return http:NO_CONTENT;
    }
}
```

#### Entity Relationship Diagram

The [entity relationship diagram view](https://wso2.com/ballerina/vscode/docs/implement-the-code/sequence-diagram-view/) for the defined data model is the following.

<img src="./resources/entity_diagram.png" alt="Entity Diagram" height="800" style="width:auto; max-width:100%">

> ### Step 3: Start the docker service

Reviewer 2: Is this for the databse server? if so let's ask the user to make sure that the databse server runs

Start the docker service using the following command.

```bash
$ docker-compose up

> ```

Reviewer: Why do we need this?

# Data integration

## Overview

> 

Reviewer: Can we describe the use case a bit? Right now, the first mention of it is in https://github.com/ballerina-guides/integration-tutorials/pull/25/files#diff-a790852c6b1441066a79743dc839c7476ca0ba55775334f9c92ab91dff226190R13 which is a bit confusing.

> In this tutorial, you will develop a service to manage data stored in a database using the Ballerina persistence layer. The service will be used to create, retrieve, update and delete data from the database.

Reviewer: ```suggestion
In this tutorial, you will develop a service to manage data stored in a database using the Ballerina persistence layer. The service will be used to create, retrieve, update, and delete data from the database.
```

To implement this use case, you will develop a REST service with multiple resources using Visual Studio Code with the Ballerina Swan Lake extension.

> Define the data model using the `bal persist` feature and generate the client objects, types and scripts for the model. Then, define the service and implement the logic to interact with the database.

Reviewer: ```suggestion
Define the data model using the `bal persist` feature and generate the client objects, types, and scripts for the model. Then, define the service and implement the logic to interact with the database.
```

        "email": "johndoe@gmail.com",
        "department": "IT"
   }
   ```

2. Send a `POST` request to the service to create a task for the employee.

    ```json
    {
        "taskId": 1001,
        "description": "Analyze network performance",
        "status": "IN_PROGRESS",
        "employeeId": 1
    }
    ```

3. Send a `PUT` request to the service to update the task status.
   
    ```json
    {
        "description": "Analyze network performance",
        "status": "COMPLETED",
        "employeeId": 1
    }
    ```

4. Delete the task by giving the task ID of the completed task that needs to be deleted.

### Concepts covered

- REST API
- The Ballerina persistence layer
- CRUD operations

## Develop the application

### Step 1: Set up the workspace

Install [Ballerina Swan Lake](https://ballerina.io/downloads/) and the [Ballerina Swan Lake extension](https://marketplace.visualstudio.com/items?itemName=wso2.ballerina) on VS Code.

### Step 2: Develop the service

Follow the instructions given in this section to develop the service.

1. Create a new Ballerina project using the `bal` command and open it in VS Code.

    ```bash
    $ bal new data-integration
    ```

> 2. Initialize `bal persist` in the project. Specify the module name as `store` and the datastore as `mysql`. This will create a persist directory in the project root directory. The directory will contain the `model.bal` file, which is used to define the data model.

Reviewer: I know the persist docs say something similar to the first sentence, but wouldn't it make sense to say something like

```
Use `bal persist` in the project root to initialize the persistence layer for the project.
```

> 2. Initialize `bal persist` in the project. Specify the module name as `store` and the datastore as `mysql`. This will create a persist directory in the project root directory. The directory will contain the `model.bal` file, which is used to define the data model.

Reviewer: a persist directory -> a directory named `persist`

    ```bash
    $ bal persist init --module store --datastore mysql
    ```

3. Open the `model.bal` file in the persist directory and define the data model as follows.

    ```ballerina
    import ballerina/persist as _;

    type Employee record {|
        readonly int id;
        string name;
        int age;
        string phone;
        string email;
        string department;
        Task[] tasks;
    |};

    type Task record {|
        readonly int taskId;
        string taskName;
        string description;
        TaskStatus status;
        Employee employee;
    |};

    public enum TaskStatus {
        NOT_STARTED,
        IN_PROGRESS,
        COMPLETED
    };
    ```

    - The above data model defines a one to many relationship, where each `Employee` can have multiple assigned `Task`s, and each `Task` is associated with only one `Employee`.   

    **Note:** 
    > The entities defined in the `model.bal` file should be based on the [`persist model specification`](https://ballerina.io/learn/persist-model/).

> 4. Generate the client objects, types and and scripts for the model using the following command. Specify `store` as the module name. This will parse the `persist/model.bal` file and add the generated files under the generated directory.

Reviewer: ```suggestion
4. Generate the client objects, types, and scripts for the model using the following command. Specify `store` as the module name. This will parse the `persist/model.bal` file and add the generated files under the generated directory.
```

> 4. Generate the client objects, types and and scripts for the model using the following command. Specify `store` as the module name. This will parse the `persist/model.bal` file and add the generated files under the generated directory.

Reviewer: This will parse the `persist/model.bal` file and add the generated files under the generated directory.

->

This will use the model defined in the `persist/model.bal` file to generate the persistence constructs in a module named `store` in a directory named `generated`.

   
    ```bash
    $ bal persist generate --module store

>     ```

Reviewer: Do we need to specify module again?

> 5. Remove the generated content in the `main.bal` file in the project root and open the diagram view in VS Code.

Reviewer: Better to include a sentence indicating that we are now going to work on the core logic using the generated constructs. Or the transition may not be obvious at first glance.

    ![Open diagram view](./resources/open_diagram_view.gif)

> 6. Define a variable of the generated client object `store:Client` type to interact with the database.

Reviewer: ```suggestion
6. Define a variable of the generated client object type (`store:Client`) to interact with the database.
```

   
   ```ballerina
   import data_integration.store;

   final store:Client dbClient = check new;
   ```
   

> 7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resources that accept user requests, interact with the database and manage operations to create, retrieve, update, and delete employee and task data.

Reviewer: ```suggestion
7. Define the [HTTP service (REST API)](https://ballerina.io/learn/by-example/#rest-service) that has the resources that accept user requests, interact with the database, and manage operations to create, retrieve, update, and delete employee and task data.
```

Check for Oxford comma usage wherever applicable.

  
   - Open the [Ballerina HTTP API Designer](https://wso2.com/ballerina/vscode/docs/design-the-services/http-api-designer) in VS Code.

   - Define a service attached to the listener that is listening on port `9090`.

        ![Define the service](./resources/define_a_service.gif)

>    - Define HTTP resources that allow CRUD operations on the employee data. The following GIF shows how to define the `POST` resource that creates an employee. Similarly, define the `GET`, `PUT` and `DELETE` resources.

Reviewer: Why are we saying "The following GIF shows..." instead of describing like we did in the other tutorials.

We basically need to have enough content for someone who just wants to follow the tutorial with written steps instead of having to rely on the GIF.

        int[]|persist:Error result = dbClient->/employees.post([employee]);
        if result is persist:AlreadyExistsError {
            return http:CONFLICT;
        }
        if result is persist:Error {
            return <http:InternalServerError>{body: result.message()};
        }
        return http:CREATED;
    }

    resource function get employees/[int empId]/tasks() returns store:Task[]|http:NotFound|http:InternalServerError {
        store:Task[]|persist:Error tasks = from store:Task task in dbClient->/tasks(store:Task)
            where task.employeeId == empId
            select task;
        if tasks is persist:NotFoundError {
            return http:NOT_FOUND;
        }
        if tasks is persist:Error {
            return <http:InternalServerError>{body: tasks.message()};
        }
        return tasks;
    }
}
```

   - The `POST` resource accepts a `store:EmployeeInsert` record as the payload and returns an `http:Created` response if the employee is created successfully. If the employee already exists, it returns an `http:Conflict` response and if an error occurs, it returns an `http:InternalServerError` response.

>    - The `GET` resource accepts an employee ID as a path parameter and retrieves tasks assigned to the employee using a database query where the `employeeId` is equal to the provided `empId`. If the employee does not exist, it returns an `http:NotFound` response and if an error occurs, it returns an `http:InternalServerError` response.

Reviewer: I would move the GET resource + logic closer to this.

        if empTask is persist:NotFoundError {
            return http:NOT_FOUND;
        }
        store:Task|persist:Error updatedTask = dbClient->/tasks/[taskId].put(task);
        if updatedTask is persist:Error {
            return <http:InternalServerError>{body: updatedTask.message()};
        }
        return updatedTask;
    }

    resource function delete tasks/[int taskId]()
            returns http:NoContent|http:NotFound|http:InternalServerError {
        store:Task|persist:Error task = dbClient->/tasks/[taskId];
        if task is persist:NotFoundError {
            return http:NOT_FOUND;
        }
        if task is persist:Error {
            return <http:InternalServerError>{body: task.message()};
        }
        if task.status != store:COMPLETED {
            return <http:InternalServerError>{body: "Task is not completed yet"};
        }
        store:Task|persist:Error deletedTask = dbClient->/tasks/[taskId].delete;
        if deletedTask is persist:Error {
            return <http:InternalServerError>{body: deletedTask.message()};
        }
        return http:NO_CONTENT;
    }
}
```

#### Entity Relationship Diagram

> 

Reviewer: Aren't we planning on including a GIF here? To show how we get the diagram?

The [entity relationship diagram view](https://wso2.com/ballerina/vscode/docs/implement-the-code/sequence-diagram-view/) for the defined data model is the following.

<img src="./resources/entity_diagram.png" alt="Entity Diagram" height="800" style="width:auto; max-width:100%">

### Step 4: Build and run the service

![Run the service](./resources/run_the_service.gif)

> **Note:**
> Alternatively, you can run this service by navigating to the project root and using the `bal run` command.
>
> ```
> data-integration$ bal run
> Compiling source
>         integration_tutorials/data_integration:0.1.0
>
> Running executable
> ```

### Step 5: Try out the use case

Let's test the use case by sending a request to the service.

#### Send a request

Use the [Try it](https://wso2.com/ballerina/vscode/docs/try-the-services/try-http-services/) feature to send a request to the service. Use the following as the JSON payload to create an employee.

```json
{
    "id": 1,
    "name": "John Doe",
    "age": 22,
    "phone": "8770586755",
    "email": "johndoe@gmail.com",
    "department": "IT"
}
```

![Send a request](./resources/try_it.gif)

> 

Reviewer: Shall we send a GET request also? To show the added employee?

# Data integration

## Overview

> In this tutorial, you will develop a service that accepts requests to create an employee and a task, update task status, and delete a completed task. Ballerina persistence layer is used to manage data stored in a database. The service will be used to create, retrieve, update, and delete data from the database.

Reviewer: Shall we have one sentence to introduce what we're working with? E.g., manage employees and tasks assigned to them.

Reviewer: Ballerina persistence layer -> The Ballerina persistence layer

> In this tutorial, you will develop a service that accepts requests to create an employee and a task, update task status, and delete a completed task. Ballerina persistence layer is used to manage data stored in a database. The service will be used to create, retrieve, update, and delete data from the database.

Reviewer: Rather than saying just "used to manage data stored in a database", may add value if we can frame it in terms of value it adds. e.g., convenience, abstracting out the underlying db layer

To implement this use case, you will develop a REST service with multiple resources using Visual Studio Code with the Ballerina Swan Lake extension.

> Define the data model using the `bal persist` feature and generate the client objects, types, and scripts for the model. Then, define the service and implement the logic to interact with the database.

Reviewer: Can we rethink this in terms of the flow. Feels disconnected after the preceding sentence.

    ```ballerina
    import ballerina/persist as _;

    type Employee record {|
        readonly int id;
        string name;
        int age;
        string phone;
        string email;
        string department;
        Task[] tasks;
    |};

    type Task record {|
        readonly int taskId;
        string taskName;
        string description;
        TaskStatus status;
        Employee employee;
    |};

    public enum TaskStatus {
        NOT_STARTED,
        IN_PROGRESS,
        COMPLETED
    };
    ```

    - The above data model defines a one to many relationship, where each `Employee` can have multiple assigned `Task`s, and each `Task` is associated with only one `Employee`.   

    **Note:** 
    > The entities defined in the `model.bal` file should be based on the [`persist model specification`](https://ballerina.io/learn/persist-model/).

4. Generate the client objects, types, and scripts for the model using the following command. This will use the model defined in the `persist/model.bal` file to generate the persistence constructs in a module named `store` in a directory named `generated`.
   
    ```bash

>     $ bal persist generate --module store

Reviewer 3: ```suggestion
    $ bal persist generate
```
We can avoid the `--module store` part in this command as it is already executed in the init command above.
